File: wmn/wmn/report/acccount_statement/acccount_statement.js
frappe.query_reports["Acccount Statement"] = {
    "filters": [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            reqd: 1,
            default: frappe.defaults.get_user_default("Company"),
            on_change: function () {
                // تحديث العملة الافتراضية عند تغيير الشركة
                let company = frappe.query_report.get_filter_value("company");
                if (company) {
                    frappe.call({
                        method: "frappe.client.get_value",
                        args: {
                            doctype: "Company",
                            fieldname: "default_currency",
                            filters: { name: company }
                        },
                        callback: function (r) {
                            if (r.message && r.message.default_currency) {
                                frappe.query_report.set_filter_value("currency", r.message.default_currency);
                            }
                        }
                    });
                }
                // تحديث قائمة الأطراف
                frappe.query_report.refresh();
            }
        },
        {
            fieldname: "party_type",
            label: __("Party Type"),
            fieldtype: "Select",
            reqd: 1,
            "default": "Customer",
            options: ["Customer", "Supplier", "Employee", "Account", "Shareholder"],
            on_change: function () {
                // مسح قيمة الطرف عند تغيير النوع
                frappe.query_report.set_filter_value("party", []);
                frappe.query_report.set_filter_value("party_name", "");
                frappe.query_report.set_filter_value("tax_id", "");
            }
        },
        {
            fieldname: "party",
            label: __("Party"),
            fieldtype: "MultiSelectList",
            reqd: 1,
            get_data: function (txt) {
                let party_type = frappe.query_report.get_filter_value("party_type");
                let company = frappe.query_report.get_filter_value("company");

                if (!party_type || !company) return [];

                if (party_type === "Account") {
                    return frappe.call({
                        method: "frappe.desk.search.search_link",
                        args: {
                            doctype: "Account",
                            txt: txt,
                            filters: {
                                "company": company,
                                "is_group": 0
                            }
                        },
                        async: false
                    }).then(r => r.message || []);
                } else {
                    return frappe.call({
                        method: "frappe.desk.search.search_link",
                        args: {
                            doctype: party_type,
                            txt: txt
                        },
                        async: false
                    }).then(r => r.message || []);
                }
            },
            on_change: function () {
                let party_type = frappe.query_report.get_filter_value("party_type");
                let parties = frappe.query_report.get_filter_value("party");
                let company = frappe.query_report.get_filter_value("company");

                if (!party_type || parties.length === 0 || !company) {
                    frappe.query_report.set_filter_value("party_name", "");
                    frappe.query_report.set_filter_value("tax_id", "");
                    return;
                }

                let party = parties[0];
                let fieldname = "name";
                if (party_type === "Customer") fieldname = "customer_name";
                else if (party_type === "Supplier") fieldname = "supplier_name";

                // تحديد الحقول المتاحة حسب نوع الطرف
                let fields_to_fetch = [fieldname];
                if (party_type === "Customer") {
                    fields_to_fetch.push("tax_id");
                } else if (party_type === "Supplier") {
                    fields_to_fetch.push("tax_id");
                }

                frappe.db.get_value(party_type, party, fields_to_fetch, function (value) {
                    frappe.query_report.set_filter_value("party_name", value[fieldname]);

                    // فقط إذا كان حقل tax_id موجوداً في نوع الطرف
                    if (fields_to_fetch.includes("tax_id")) {
                        frappe.query_report.set_filter_value("tax_id", value["tax_id"] || "");
                    } else {
                        frappe.query_report.set_filter_value("tax_id", "");
                    }

                    if (parties.length > 1) {
                        // الاحتفاظ بالعنصر الأخير فقط
                        let last_party = parties[parties.length - 1];
                        frappe.query_report.set_filter_value("party", [last_party]);
                        parties = [last_party];
                    }

                    // تحديث التقرير بعد تعيين القيم
                    frappe.query_report.refresh();
                });
            }
        },
        {
            fieldname: "party_name",
            label: __("Party Name"),
            fieldtype: "Data",
            hidden: 1,
            read_only: 1
        },
        {
            fieldname: "tax_id",
            label: __("Tax Id"),
            fieldtype: "Data",
            hidden: 1,
            read_only: 1
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            "default": frappe.datetime.get_today(),
            reqd: 1
        },
        {
            "fieldname": "use_transaction_currency",
            "label": "عرض بالعملة",
            "fieldtype": "Check",
            "default": 0
        },
        {
            "fieldname": "currency",
            "label": "currency",
            "fieldtype": "Link",
            "options": "Currency",
            "default": "SAR"
        }
    ],


    onload: function (report) {
        report.page.set_primary_action(__('Print'), function () {
            let data = report.data;

            if (!data || data.length === 0) {
                frappe.msgprint("لا توجد بيانات للطباعة.");
                return;
            }

            let party = report.get_values().party || "العميل";
            let from_date = report.get_values().from_date || "";
            let to_date = report.get_values().to_date || "";

            // جلب شعار الشركة بشكل صحيح باستخدام Promise
            let company_name = report.get_values().company;
            let company_logo = "/files/company_logo.png"; // شعار افتراضي

            // استخدام Promise للتعامل مع الطلب غير المتزامن
            frappe.call({
                method: "frappe.client.get",
                args: {
                    doctype: "Company",
                    name: company_name
                },
                callback: function (r) {
                    if (r.message && r.message.company_logo) {
                        company_logo = r.message.company_logo;
                    }

                    // متابعة عملية الطباعة بعد الحصول على الشعار
                    printReport(data, party, from_date, to_date, company_logo);
                }
            });
        });

        // دالة منفصلة للطباعة
        function printReport(data, party, from_date, to_date, company_logo) {
            let report_title = `كشف حركة حساب : ${party}`;
            let doc_type_map = {
                "Sales Invoice": "فاتورة مبيعات",
                "Purchase Invoice": "فاتورة مشتريات",
                "Journal Entry": "قيد محاسبي",
                "Pay": "سند صرف",
                "Receive": "سند قبض",
                "No Remarks": " "
            };

            // إنشاء iframe للطباعة
            let iframe = document.createElement('iframe');
            iframe.style.position = 'absolute';
            iframe.style.width = '0';
            iframe.style.height = '0';
            iframe.style.border = 'none';
            document.body.appendChild(iframe);

            let doc = iframe.contentWindow.document;

            // عدد الصفوف في كل صفحة
            const rowsPerPage = 25;
            const totalPages = Math.ceil(data.length / rowsPerPage);

            doc.open();
            doc.write(`
                <!DOCTYPE html>
                <html dir="rtl">
                <head>
                    <meta charset="UTF-8">
                    <title>${report_title}</title>
                    <style>
                        body {
                            font-family: Arial, sans-serif;
                            margin: 0;
                            padding: 0;
                            direction: rtl;
                            width: 100%;
                        }
                        .header {
                            display: flex;
                            justify-content: space-between;
                            align-items: center;
                            margin-bottom: 15px;
                            padding-bottom: 10px;
                            border-bottom: 2px solid #090960;
                        }
                        table {
                            width: 99%;
                            border-collapse: collapse;
                            margin: 0 auto 15px auto;
                            font-size: 15px;
                            table-layout: fixed;
                        }
                        th, td {
                            border: 1px solid #000;
                            padding: 8px 5px;
                            text-align: center;
                            word-wrap: break-word;
                        }
                        th {
                            background-color: #e8ecf7;
                            font-weight: bold;
                            font-size: 16px;
                            padding: 10px 5px;
                        }
                        tr:nth-child(even) {
                            background-color: #f8f8f8;
                        }

                        /* التحكم في عرض الأعمدة بالنسبة المئوية */
                        th:nth-child(1), td:nth-child(1) { width: 10%; } /* التاريخ */
                        th:nth-child(2), td:nth-child(2) { width: 10%; } /* المستند */
                        th:nth-child(3), td:nth-child(3) { width: 10%; } /* الرقم */
                        th:nth-child(4), td:nth-child(4) { width: 35%; text-align: right; padding-right: 14px; } /* الملاحظات */
                        th:nth-child(5), td:nth-child(5) { width: 10%; font-size: 15px; } /* مدين */
                        th:nth-child(6), td:nth-child(6) { width: 10%; font-size: 15px; } /* دائن */
                        th:nth-child(7), td:nth-child(7) { width: 15%; font-size: 15px; } /* الرصيد */

                        .currency {
                            font-size: 9px;
                            color: #666;
                            margin-left: 3px;
                            font-weight: normal;
                        }

                        .amount {
                            font-size: 14px;
                            font-weight: bold;
                        }

                        .footer {
                            margin-top: 20px;
                            text-align: center;
                            font-size: 12px;
                            color: #666;
                            width: 100%;
                            padding-top: 10px;
                            border-top: 1px solid #ddd;
                        }

                        .page-break {
                            page-break-after: always;
                            break-after: page;
                        }

                        .pagination {
                            text-align: center;
                            margin: 10px 0;
                            font-size: 12px;
                            color: #666;
                        }

                        @media print {
                            @page {
                                margin: 0.7cm;
                                size: landscape;
                            }
                            body {
                                margin: 0;
                                padding: 0;
                                width: 100%;
                            }
                            .header {
                                position: running(header);
                            }
                            .footer {
                                position: running(footer);
                            }
                            table {
                                width: 100%;
                                margin: 0 auto 10px auto;
                            }
                            th {
                                background-color: #e8ecf7 !important;
                                -webkit-print-color-adjust: exact;
                                print-color-adjust: exact;
                                font-size: 16px !important;
                            }
                            tr:nth-child(even) {
                                background-color: #f8f8f8 !important;
                                -webkit-print-color-adjust: exact;
                                print-color-adjust: exact;
                            }
                            /* التأكد من أن الأعمدة تحافظ على نسبها أثناء الطباعة */
                            th:nth-child(1), td:nth-child(1) { width: 10% !important; }
                            th:nth-child(2), td:nth-child(2) { width: 10% !important; }
                            th:nth-child(3), td:nth-child(3) { width: 10% !important; }
                            th:nth-child(4), td:nth-child(4) { width: 35% !important; text-align: right !important; padding-right: 15px !important; }
                            th:nth-child(5), td:nth-child(5) { width: 10% !important; font-size: 15px !important; }
                            th:nth-child(6), td:nth-child(6) { width: 10% !important; font-size: 15px !important; }
                            th:nth-child(7), td:nth-child(7) { width: 15% !important; font-size: 15px !important; }

                            .currency {
                                font-size: 9px !important;
                            }

                            .amount {
                                font-size: 14px !important;
                            }

                            .page-break {
                                page-break-after: always !important;
                                break-after: page !important;
                            }
                        }
                    </style>
                </head>
                <body>
            `);

            // تقسيم البيانات إلى صفحات
            doc.write(`
    <div class="header">
        <div style="width:30%;">
            <img src="${company_logo}" style="max-height:60px;">
        </div>
        <div style="width:50%; text-align:center;">
            <h1 style="font-weight:bold; font-size:18px; color:#090960; margin:0;">
                ${report_title}
            </h1>
        </div>
        <div style="width:20%; text-align:right; color:#090960; font-size:13px;">
            <div style="margin-bottom:5px;">
                <strong>من:</strong> ${from_date}
            </div>
            <div>
                <strong>إلى:</strong> ${to_date}
            </div>
        </div>
    </div>

    <table>
        <thead>
            <tr>
                <th>التاريخ</th>
                <th>المستند</th>
                <th>الرقم</th>
                <th>ملاحظات</th>
                <th>مدين</th>
                <th>دائن</th>
                <th>الرصيد</th>
            </tr>
        </thead>
        <tbody>
`);

            data.forEach((row, idx) => {
                let currency = row.currency || "YER";
                doc.write(`
        <tr style="background-color: ${idx % 2 === 0 ? '#F8F8F8' : '#FFFFFF'};">
            <td>${row.posting_date}</td>
            <td>${doc_type_map[row.voucher_subtype] || row.voucher_subtype}</td>
            <td>${row.voucher_no}</td>
            <td style="text-align:right; padding-right:15px;">${row.remarks || ""}</td>
            <td style="text-align:right;">
                <span class="amount">${parseFloat(row.debit).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}</span>
                <span class="currency">${currency}</span>
            </td>
            <td style="text-align:right;">
                <span class="amount">${parseFloat(row.credit).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}</span>
                <span class="currency">${currency}</span>
            </td>
            <td style="text-align:right;">
                <span class ="amount">${parseFloat(row.balance).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2})}</span>
                <span class="currency">${currency}</span>
            </td>
        </tr>
    `);
            });

            doc.write(`
        </tbody>
    </table>

    <div class="footer">
        المحاسب                    المراجع                        المدير العام
    </div>
`);
            doc.close();

            // الانتظار حتى يتم تحميل المحتوى ثم الطباعة
            setTimeout(function () {
                iframe.contentWindow.focus();
                iframe.contentWindow.onafterprint = function () {
                    document.body.removeChild(iframe);
                };
                iframe.contentWindow.print();
                setTimeout(function () {
                    if (document.body.contains(iframe)) {
                        document.body.removeChild(iframe);
                    }
                }, 1000);
            }, 500);
        }
    }
};

File: wmn/wmn/report/acccount_statement/acccount_statement.py
import frappe
from frappe import _

def execute(filters=None):
    if not filters:
        filters = {}
    
    columns = get_columns(filters)
    data = get_data(filters)
    
    return columns, data

def get_columns(filters):
    """إرجاع الأعمدة بناءً على الفلاتر"""
    base_columns = [
        {"label": "التاريخ", "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
        {"label": "نوع السند", "fieldname": "voucher_subtype", "fieldtype": "Data", "width": 150},
        {"label": "رقم السند", "fieldname": "voucher_no", "fieldtype": "Dynamic Link", "options": "voucher_type", "width": 180},
        {"label": "ملاحظات", "fieldname": "remarks", "fieldtype": "Data", "width": 200},
    ]
    
    use_transaction_currency = filters.get("use_transaction_currency")
    
    if use_transaction_currency:
        currency = filters.get("currency", "YER")
        columns = base_columns + [
            {"label": "سعر الصرف", "fieldname": "exchange_rate", "fieldtype": "Float", "width": 100, "precision": 6},
            {"label": "مدين", "fieldname": "debit", "fieldtype": "Currency", "options": "currency", "width": 120},
            {"label": "دائن", "fieldname": "credit", "fieldtype": "Currency", "options": "currency", "width": 120},
            {"label": "الرصيد", "fieldname": "balance", "fieldtype": "Currency", "options": "currency", "width": 120},
            {"label": "currency", "fieldname": "currency", "fieldtype": "Data", "hidden": 1},
        ]
    else:
        columns = base_columns + [
            {"label": "مدين", "fieldname": "debit", "fieldtype": "Currency", "width": 120},
            {"label": "دائن", "fieldname": "credit", "fieldtype": "Currency", "width": 120},
            {"label": "الرصيد", "fieldname": "balance", "fieldtype": "Currency", "width": 120},
        ]
    
    return columns

def get_filters(filters):
    """بناء فلتر Script Report"""
    return [
        {
            "fieldname": "company",
            "label": _("الشركة"),
            "fieldtype": "Link",
            "options": "Company",
            "reqd": 1,
            "default": frappe.defaults.get_user_default("Company")
        },
        {
            "fieldname": "party_type",
            "label": _("نوع الطرف"),
            "fieldtype": "Select",
            "options": "\nCustomer\nSupplier\nEmployee\nAccount",
            "reqd": 1
        },
        {
            "fieldname": "party",
            "label": _("الطرف"),
            "fieldtype": "MultiSelectList",
            "get_data": get_party_options,
            "reqd": 1
        },
        {
            "fieldname": "from_date",
            "label": _("من تاريخ"),
            "fieldtype": "Date",
            "reqd": 1,
            "default": frappe.utils.nowdate()
        },
        {
            "fieldname": "to_date",
            "label": _("إلى تاريخ"),
            "fieldtype": "Date",
            "reqd": 1,
            "default": frappe.utils.nowdate()
        },
        {
            "fieldname": "use_transaction_currency",
            "label": _("عرض بعملة أخرى"),
            "fieldtype": "Check",
            "default": 0
        },
        {
            "fieldname": "currency",
            "label": _("العملة"),
            "fieldtype": "Link",
            "options": "Currency",
            "depends_on": "use_transaction_currency",
            "mandatory_depends_on": "use_transaction_currency",
            "default": frappe.db.get_value("Company", frappe.defaults.get_user_default("Company"), "default_currency")
        }
    ]

def get_party_options(doctype, txt, searchfield, start, page_len, filters):
    """جلب قائمة الأطراف بناءً على نوع الطرف المحدد"""
    party_type = filters.get("party_type")
    
    if not party_type:
        return []
    
    if party_type == "Account":
        return frappe.db.sql("""
            SELECT name, name 
            FROM `tabAccount`
            WHERE is_group = 0
              AND (name LIKE %s OR account_name LIKE %s)
              AND company = %s
            ORDER BY name
            LIMIT %s, %s
        """, (f"%{txt}%", f"%{txt}%", filters.get("company"), start, page_len))
    
    else:
        # لـ Customer, Supplier, Employee
        party_doctype_map = {
            "Customer": "Customer",
            "Supplier": "Supplier",
            "Employee": "Employee"
        }
        
        party_doctype = party_doctype_map.get(party_type)
        if not party_doctype:
            return []
        
        return frappe.db.sql("""
            SELECT name, name 
            FROM `tab{0}`
            WHERE name LIKE %s 
               OR customer_name LIKE %s
            ORDER BY name
            LIMIT %s, %s
        """.format(party_doctype), (f"%{txt}%", f"%{txt}%", start, page_len))

def validate_filters(filters):
    """التحقق من صحة الفلاتر"""
    if not filters.get("company"):
        frappe.throw(_("الرجاء تحديد الشركة"))
    
    if not filters.get("party_type"):
        frappe.throw(_("الرجاء تحديد نوع الطرف"))
    
    if not filters.get("party") or len(filters.get("party", [])) == 0:
        frappe.throw(_("الرجاء تحديد طرف واحد على الأقل"))
    
    if len(filters.get("party", [])) > 1:
        frappe.throw(_("الرجاء اختيار طرف واحد فقط للعرض في التقرير"))

def get_exchange_rate_simple(from_currency, to_currency, date, company):
    """نسخة مبسطة من دالة سعر الصرف مع مراعاة الشركة"""
    if from_currency == to_currency:
        return 1.0
    
    try:
        # جلب سعر الصرف من جدول Currency Exchange مع مراعاة الشركة
        rate = frappe.db.get_value(
            "Currency Exchange",
            filters={
                "date": ["<=", date],
                "from_currency": from_currency,
                "to_currency": to_currency
            },
            fieldname="exchange_rate",
            order_by="date desc"
        )
        
        if rate:
            return float(rate)
        
        # محاولة العكس
        rate = frappe.db.get_value(
            "Currency Exchange",
            filters={
                "date": ["<=", date],
                "from_currency": to_currency,
                "to_currency": from_currency
            },
            fieldname="exchange_rate",
            order_by="date desc"
        )
        
        if rate:
            return 1.0 / float(rate)
            
    except Exception:
        pass
    
    # السعر الافتراضي
    frappe.msgprint(_("تحذير: لم يتم العثور على سعر صرف لـ {0} إلى {1} بتاريخ {2}").format(
        from_currency, to_currency, date
    ))
    return 1.0

def get_previous_balance(party_type, party, from_date, company, company_currency, selected_currency):
    """حساب الرصيد السابق مع مراعاة الشركة"""
    # بناء الفلاتر
    filters = {
        "posting_date": ["<", from_date],
        "is_cancelled": 0,
        "company": company
    }
    
    if party_type == "Account":
        filters["account"] = party
    else:
        filters.update({
            "party_type": party_type,
            "party": party
        })
    
    # جلب البيانات
    gl_entries = frappe.get_all(
        "GL Entry",
        fields=[
            "posting_date",
            "debit", "credit",
            "debit_in_transaction_currency", 
            "credit_in_transaction_currency",
            "transaction_currency"
        ],
        filters=filters
    )
    
    balance = 0.0
    
    for entry in gl_entries:
        if (entry.transaction_currency == selected_currency and 
            (entry.debit_in_transaction_currency or 0) + (entry.credit_in_transaction_currency or 0) > 0):
            debit = entry.debit_in_transaction_currency or 0
            credit = entry.credit_in_transaction_currency or 0
            rate = 1.0
        else:
            debit = entry.debit or 0
            credit = entry.credit or 0
            rate = get_exchange_rate_simple(
                company_currency, 
                selected_currency, 
                entry.posting_date,
                company
            ) or 1.0
        
        balance += (debit * rate) - (credit * rate)
    
    return balance

def get_data(filters):
    """جلب البيانات الرئيسية للتقرير"""
    validate_filters(filters)
    
    company = filters.get("company")
    party_type = filters.get("party_type")
    party = filters.get("party")[0]  # نأخذ الطرف الأول فقط
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    use_transaction_currency = filters.get("use_transaction_currency", False)
    selected_currency = filters.get("currency")
    
    # الحصول على العملة الأساسية للشركة
    company_currency = frappe.db.get_value("Company", company, "default_currency") or "YER"
    
    if not selected_currency:
        selected_currency = company_currency
    
    result = []
    
    # 1. حساب الرصيد السابق
    previous_balance = 0
    
    if use_transaction_currency and selected_currency != company_currency:
        previous_balance = get_previous_balance(
            party_type, party, from_date, company, company_currency, selected_currency
        )
    else:
        # حساب الرصيد باستخدام العملة الأساسية
        if party_type == "Account":
            prev_data = frappe.db.sql("""
                SELECT COALESCE(SUM(debit - credit), 0) as balance
                FROM `tabGL Entry`
                WHERE account = %s 
                  AND company = %s
                  AND is_cancelled = 0
                  AND posting_date < %s
            """, (party, company, from_date), as_dict=True)
        else:
            prev_data = frappe.db.sql("""
                SELECT COALESCE(SUM(debit - credit), 0) as balance
                FROM `tabGL Entry`
                WHERE party_type = %s
                  AND party = %s
                  AND company = %s
                  AND is_cancelled = 0
                  AND posting_date < %s
            """, (party_type, party, company, from_date), as_dict=True)
        
        previous_balance = prev_data[0].balance if prev_data else 0
    
    # إضافة سطر الرصيد السابق
    result.append({
        "posting_date": from_date,
        "voucher_subtype": "رصيد سابق",
        "voucher_no": "",
        "debit": 0,
        "credit": 0,
        "remarks": "",
        "balance": previous_balance,
        "currency": selected_currency if use_transaction_currency else company_currency,
        "exchange_rate": 1.0,
        "voucher_type": ""
    })
    
    # 2. جلب الحركات في الفترة
    if party_type == "Account":
        entries_query = """
            SELECT 
                posting_date,
                voucher_subtype,
                voucher_no,
                voucher_type,
                remarks,
                SUM(debit) as debit,
                SUM(credit) as credit,
                SUM(debit_in_transaction_currency) as debit_in_transaction_currency,
                SUM(credit_in_transaction_currency) as credit_in_transaction_currency,
                AVG(transaction_exchange_rate) as transaction_exchange_rate,
                MAX(transaction_currency) as transaction_currency
            FROM `tabGL Entry`
            WHERE account = %s
              AND company = %s
              AND is_cancelled = 0
              AND posting_date BETWEEN %s AND %s
            GROUP BY posting_date, voucher_subtype, voucher_no, voucher_type, remarks
            ORDER BY posting_date ASC, voucher_no ASC
        """
        params = (party, company, from_date, to_date)
    else:
        entries_query = """
            SELECT 
                posting_date,
                voucher_subtype,
                voucher_no,
                voucher_type,
                remarks,
                SUM(debit) as debit,
                SUM(credit) as credit,
                SUM(debit_in_transaction_currency) as debit_in_transaction_currency,
                SUM(credit_in_transaction_currency) as credit_in_transaction_currency,
                AVG(transaction_exchange_rate) as transaction_exchange_rate,
                MAX(transaction_currency) as transaction_currency
            FROM `tabGL Entry`
            WHERE party_type = %s
              AND party = %s
              AND company = %s
              AND is_cancelled = 0
              AND posting_date BETWEEN %s AND %s
            GROUP BY posting_date, voucher_subtype, voucher_no, voucher_type, remarks
            ORDER BY posting_date ASC, voucher_no ASC
        """
        params = (party_type, party, company, from_date, to_date)
    
    entries = frappe.db.sql(entries_query, params, as_dict=True)
    
    # 3. معالجة الحركات وإضافة الرصيد الجاري
    running_balance = previous_balance
    type_map = {
        "Pay": "سند صرف",
        "Receive": "سند قبض",
        "Sales Invoice": "فاتورة مبيعات",
        "Purchase Invoice": "فاتورة مشتريات",
        "Journal Entry": "قيد محاسبي",
        "Payment Entry": "سند دفع/قبض",
        "Expense Claim": "مطالبة مصروفات",
        "Salary Slip": "مسير رواتب"
    }
    
    for entry in entries:
        exchange_rate = 1.0
        debit = entry.debit or 0
        credit = entry.credit or 0
        
        if use_transaction_currency and selected_currency != company_currency:
            if (entry.transaction_currency == selected_currency and
                entry.transaction_exchange_rate and
                (entry.debit_in_transaction_currency or 0) + (entry.credit_in_transaction_currency or 0) > 0):
                
                debit = entry.debit_in_transaction_currency or 0
                credit = entry.credit_in_transaction_currency or 0
                exchange_rate = 1.0
            else:
                exchange_rate = get_exchange_rate_simple(
                    company_currency, 
                    selected_currency, 
                    entry.posting_date,
                    company
                )
                debit = debit * exchange_rate
                credit = credit * exchange_rate
        
        running_balance += debit - credit
        
        result.append({
            "posting_date": entry.posting_date,
            "voucher_subtype": type_map.get(entry.voucher_subtype, entry.voucher_subtype),
            "voucher_no": entry.voucher_no,
            "voucher_type": entry.voucher_type,
            "remarks": entry.remarks or "",
            "debit": debit,
            "credit": credit,
            "balance": running_balance,
            "currency": selected_currency if use_transaction_currency else company_currency,
            "exchange_rate": exchange_rate
        })
    
    return result

def get_chart_data(data, filters):
    """إنشاء رسم بياني للبيانات"""
    if not data or len(data) <= 1:
        return None
    
    company = filters.get("company")
    party_type = filters.get("party_type")
    party = filters.get("party")[0] if filters.get("party") else ""
    
    chart = {
        "data": {
            "labels": [d.get("posting_date") for d in data if d.get("posting_date")],
            "datasets": [
                {
                    "name": _("الرصيد"),
                    "values": [d.get("balance") for d in data if d.get("balance")],
                    "chartType": "line"
                }
            ]
        },
        "type": "line",
        "title": _("تقرير حركات الطرف: {0} - {1}").format(party_type, party),
        "height": 300
    }
    
    return chart